core-datastore/src/main/java/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt

```diff
@@ -106,51 +120,47 @@ class NiaPreferencesDataSource @Inject constructor(
         }
     }
 
-    suspend fun setFollowedAuthorIds(followedAuthorIds: Set<String>) {
-        try {
-            userPreferences.updateData {
-                it.copy {
-                    this.followedAuthorIds.clear()
-                    this.followedAuthorIds.addAll(followedAuthorIds)
-                }
-            }
-        } catch (ioException: IOException) {
-            Log.e("NiaPreferences", "Failed to update user preferences", ioException)
-        }
+    /**
+     * Adds or removes [value] from the [DslList] provided by [listGetter]
+     */
+    private suspend fun <T : DslProxy> DataStore<UserPreferences>.editList(
+        add: Boolean,
+        value: String,
+        listGetter: (UserPreferencesKt.Dsl) -> DslList<String, T>,
+        clear: UserPreferencesKt.Dsl.(DslList<String, T>) -> Unit,
+        addAll: UserPreferencesKt.Dsl.(DslList<String, T>, Iterable<String>) -> Unit
+    ) {
+        setList(
+            listGetter = listGetter,
+            listSetter = { currentList ->
+                if (add) currentList + value
+                else currentList - value
+            },
+            clear = clear,
+            addAll = addAll
+        )
     }
 
-    suspend fun toggleFollowedAuthorId(followedAuthorId: String, followed: Boolean) {
+    /**
+     * Sets the value provided by [listSetter] into the [DslList] read by [listGetter]
+     */
+    private suspend fun <T : DslProxy> DataStore<UserPreferences>.setList(
+        listGetter: (UserPreferencesKt.Dsl) -> DslList<String, T>,
+        listSetter: (DslList<String, T>) -> List<String>,
```

Comment:
It seems like there's maybe a little too much abstraction here?

It's a bit confusing that `listSetter` doesn't actually set the list passed to it, but it (may) modify the list it is given, and returns the list that should later be added.

I get why you're doing this though, since the Kotlin bindings that are automatically generated this less than ideal to do.

I think the original way without the `editList`/`setList` helpers was fine, and more straightforward, if a bit verbose.

Maybe we can add a feature request to the protobuf Kotlin bindings to allow exposing these as `MutableList`s?

Reply:
Great point on the complexity of the abstraction, but it felt clumsy and error prone copy pasting the same try block 5 times with minimum variation, especially amongst the toggle methods for topics, authors and news resources.

I've renamed `listSetter` to `listModifier` to better convey what it does.